packages/server/src/fhir/lookups/lookuptable.ts

```diff
@@ -47,13 +51,64 @@ export abstract class LookupTable {
    */
   abstract isIndexed(searchParam: SearchParameter, resourceType: string): boolean;
 
+  /**
+   * Extracts the specific values to be indexed from a resource for this table.
+   * @param result - The array that rows to be inserted should be added to.
+   * @param resource - The resource to extract values from.
+   */
+  protected abstract extractValues(result: LookupTableRow[], resource: WithId<Resource>): void;
+
   /**
    * Indexes the resource in the lookup table.
    * @param client - The database client.
    * @param resource - The resource to index.
    * @param create - True if the resource should be created (vs updated).
+   * @returns Promise on completion.
    */
-  abstract indexResource(client: PoolClient, resource: WithId<Resource>, create: boolean): Promise<void>;
+  indexResource(client: PoolClient, resource: WithId<Resource>, create: boolean): Promise<void> {
+    return this.batchIndexResources(client, [resource], create);
+  }
+
+  /**
+   * Indexes the resource in the lookup table.
+   * @param client - The database client.
+   * @param resources - The resources to index.
+   * @param create - True if the resource should be created (vs updated).
+   */
+  async batchIndexResources<T extends Resource>(
+    client: PoolClient,
+    resources: WithId<T>[],
+    create: boolean
+  ): Promise<void> {
+    if (resources.length === 0) {
+      return;
+    }
+
+    const resourceType = resources[0].resourceType;
+
+    if (!create) {
+      await this.batchDeleteValuesForResources(client, resources);
+    }
+
+    const resourceBatchSize = 500;
+    for (let i = 0; i < resources.length; i += resourceBatchSize) {
+      const batch = resources.slice(i, i + resourceBatchSize);
```

Comment:
Do we need a slice here if we're just going to iterate over a segment of the resources array and create new rows from that segment?

Reply:
The slicing at the resource level was intended to avoid tying up the event loop for too long with synchronous work without any breaks between the calls to `await this.insertValuesForResource(client, resourceType, newRows)` below. 500 was an arbitrary choice and seems a bit aggressive now that I think about it.

Reply:
The conceptual "slicing" is fine but we don't need to create an intermediate array do we?

Instead of `for (const resource of batch) {`

We could do
```
  for (let j = i; j < i + resourceBatchSize; j++) {
    resource = resources[j];
  }
```

Minor optimization but we would create a lot less intermediate arrays which have to get GC'd in quick succession

Reply:
Ah, sorry, I misunderstood your comment. Looking at it again